clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiLookupStrategy.java

```diff
@@ -0,0 +1,94 @@
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.OptionalInt;
+import java.util.Set;
+
+public interface AdminApiLookupStrategy<T> {
+
+    /**
+     * Define the scope of a given key for lookup. Key lookups are complicated
+     * by the need to accommodate different batching mechanics. For example,
+     * a `Metadata` request supports arbitrary batching of topic partitions in
+     * order to discover partitions leaders. This can be supported by returning
+     * a single scope object for all keys.
+     *
+     * On the other hand, `FindCoordinator` request only supports lookup of a
+     * single key. This can be supported by returning a different scope object
+     * for each lookup key.
+     *
+     * @param key the lookup key
+     *
+     * @return request scope indicating how lookup requests can be batched together
+     */
+    RequestScope lookupScope(T key);
+
+    /**
+     * Build the lookup request for a set of keys. The grouping of the keys is controlled
+     * through {@link #lookupScope(Object)}. In other words, each set of keys that map
+     * to the same request scope object will be sent to this method.
+     *
+     * @param keys the set of keys that require lookup
+     *
+     * @return a builder for the lookup request
+     */
+    AbstractRequest.Builder<?> buildRequest(Set<T> keys);
+
+    /**
+     * Callback that is invoked when a lookup request returns successfully. The handler
+     * should parse the response, check for errors, and return a result indicating
+     * which keys were mapped to a brokerId successfully and which keys received
+     * a fatal error (e.g. a topic authorization failure).
+     *
+     * Note that keys which receive a retriable error should be left out of the
+     * result. They will be retried automatically. For example, if the response of
+     * `FindCoordinator` request indicates an unavailable coordinator, then the key
+     * should be left out of the result so that the request will be retried.
+     *
+     * @param keys the set of keys from the associated request
+     * @param response the response received from the broker
+     *
+     * @return a result indicating which keys mapped successfully to a brokerId and
+     *         which encountered a fatal error
+     */
+    LookupResult<T> handleResponse(Set<T> keys, AbstractResponse response);
+
+    class LookupResult<K> {
+        public final Map<K, Integer> mappedKeys;
+        public final Map<K, Throwable> failedKeys;
+
+        public LookupResult(
+            Map<K, Throwable> failedKeys,
+            Map<K, Integer> mappedKeys
+        ) {
+            this.failedKeys = Collections.unmodifiableMap(failedKeys);
+            this.mappedKeys = Collections.unmodifiableMap(mappedKeys);
+        }
+    }
+
+    interface RequestScope {
+        default OptionalInt destinationBrokerId() {
+            return OptionalInt.empty();
+        }
+    }
```

Comment:
Have you decided to put this interface here because it is mainly used by the `AdminApiLookupStrategy`? Intuitively, I would have placed it in the driver because it is used by few other classes as well.

Reply:
Let me think about it. I wasn't entirely happy with this location either.

Reply:
I decided to pull this up to the top level. Let me know if that seems ok.